src/extension/vscode/client.ts

```diff
@@ -0,0 +1,162 @@
+  url: string | URL
+): {
+  connected: Promise<void>;
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const clientRef = new WeakRef(client);
+  const registered = _registerClient(clientRef, url);
+  const registry = new FinalizationRegistry(registered.unregister);
+  const unregisterToken = {};
+  registry.register(client, "CLIENT_GC", unregisterToken);
+  registered.onCleanup(() => registry.unregister(unregisterToken));
+  return registered;
+}
+
+function makeErrorHandler(cleanup: (reason?: Reason) => void): EventListener {
+  return (e) => {
+    console.error("Apollo DevTools Client encountered WebSocket error:", e);
+    cleanup("WS_ERROR");
+  };
+}
+
+export function _registerClient(
+  clientRef: WeakRef<ApolloClient<any>>,
+  url: string | URL
+): {
+  connected: Promise<void>;
+  unregister: (reason?: Reason) => void;
+  onCleanup: (cleanup: (reason: Reason) => void) => void;
+} {
+  const { signal, cleanup, registerCleanup } = getCleanupController();
+
+  function getClient() {
+    const client = clientRef.deref();
+    if (client) {
+      return getPrivateAccess(client);
+    } else {
+      cleanup("CLIENT_GC");
+    }
+  }
+
+  const ws = new WebSocket(url);
+  registerCleanup(ws.close.bind(ws, 1000));
+  ws.addEventListener("close", cleanup.bind(undefined, "WS_DISCONNECTED"), {
+    once: true,
+    signal,
+  });
+  ws.addEventListener("error", makeErrorHandler(cleanup), {
+    once: true,
+    signal,
+  });
+
+  const wsAdapter = createWsAdapter(ws, signal);
+  const wsRpcClient = createRpcClient(wsAdapter);
+  const wsRpcHandler = createRpcHandler(wsAdapter);
+  const wsActor = createActor(wsAdapter);
+  function getQueriesForClient() {
+    return getQueries(
+      getClient()?.queryManager.getObservableQueries("active") ?? new Map()
+    );
+  }
+  function getMutationsForClient() {
+    return getMutations(getClient()?.queryManager.mutationStore ?? {});
+  }
+  registerCleanup(
+    wsRpcHandler("getQueries", getQueriesForClient),
+    wsRpcHandler("getMutations", getMutationsForClient),
+    wsRpcHandler("getCache", () => getClient()?.cache.extract(true) ?? {}),
+    handleExplorerRequests(wsActor, getClient)
+  );
+
+  ws.addEventListener(
+    "open",
+    function open() {
+      const client = getClient();
+      if (!client) {
+        return;
+      }
+      wsActor.send({
+        type: "registerClient",
+        payload: {
+          id: "",
+          name:
+            "devtoolsConfig" in client ? client.devtoolsConfig.name : undefined,
+          version: client.version,
+          queryCount: getQueriesForClient().length,
+          mutationCount: getMutationsForClient().length,
+        },
+      });
+      loadErrorCodes(wsRpcClient, client.version);
+    },
+    { once: true, signal }
+  );
+
+  return {
+    connected: new Promise<void>((resolve) =>
+      ws.addEventListener("open", () => resolve(), { once: true, signal })
+    ),
```

Comment:
I'm not a huge fan of this name mostly because reading this I keep expecting this to be a boolean, but its a promise instead.

Out of curiosity, could we just use a callback here and call this `onConnected`? We can remove the promise altogether then. 

```suggestion
    onConnected: (callback: () => void) => {
      ws.addEventListener("open", () => callback(), { once: true, signal })
    },
```

Reply:
That would miss the event if you added an `onConnected` event after the connection is already established and handling *that* would add extra logic. I'm open to changing the name, but I'd like to keep the promise, as that handles that nicely.

Reply:
Thats a good point. Let's see if we can rename this then. 

Would it be out of the question for `registerClient` (or whatever the new name will be) to return a promise instead? If not, my brain is out of names 😆

Reply:
`connectedPromise` - will at least not be confused with a boolean, and also rejects now when there is a connection error